File: frontend/src/test-support/workCode.ts
export const syntheticWorkCodePrefixes = ["RJ", "BJ", "VJ", "CC"] as const;

export type SyntheticWorkCodePrefix = (typeof syntheticWorkCodePrefixes)[number];

const workCodesPerPrefix = 100;

export function syntheticWorkCode(prefix: SyntheticWorkCodePrefix, ordinal: number) {
  if (!syntheticWorkCodePrefixes.includes(prefix)) {
    throw new RangeError(`Unsupported synthetic work-code prefix: ${prefix}`);
  }
  if (!Number.isInteger(ordinal) || ordinal < 0 || ordinal >= workCodesPerPrefix) {
    throw new RangeError(`Synthetic work-code ordinal is outside 0..99: ${ordinal}`);
  }
  return `${prefix}${String(ordinal).padStart(8, "0")}`;
}

export function syntheticWorkCodeAt(index: number) {
  if (!Number.isInteger(index) || index < 0 || index >= syntheticWorkCodePrefixes.length * workCodesPerPrefix) {
    throw new RangeError(`Synthetic work-code index is outside 0..399: ${index}`);
  }
  return syntheticWorkCode(
    syntheticWorkCodePrefixes[Math.floor(index / workCodesPerPrefix)],
    index % workCodesPerPrefix,
  );
}

File: backend/internal/testfixture/work_code.go
package testfixture

import "fmt"

type WorkCodePrefix string

const (
	PrefixRJ WorkCodePrefix = "RJ"
	PrefixBJ WorkCodePrefix = "BJ"
	PrefixVJ WorkCodePrefix = "VJ"
	PrefixCC WorkCodePrefix = "CC"

	workCodesPerPrefix = 100
)

var workCodePrefixes = [...]WorkCodePrefix{PrefixRJ, PrefixBJ, PrefixVJ, PrefixCC}

func WorkCode(prefix WorkCodePrefix, ordinal int) string {
	if !validWorkCodePrefix(prefix) {
		panic(fmt.Sprintf("unsupported synthetic work-code prefix %q", prefix))
	}
	if ordinal < 0 || ordinal >= workCodesPerPrefix {
		panic(fmt.Sprintf("synthetic work-code ordinal %d is outside 0..99", ordinal))
	}
	return fmt.Sprintf("%s%08d", prefix, ordinal)
}

func WorkCodeAt(index int) string {
	if index < 0 || index >= len(workCodePrefixes)*workCodesPerPrefix {
		panic(fmt.Sprintf("synthetic work-code index %d is outside 0..399", index))
	}
	return WorkCode(workCodePrefixes[index/workCodesPerPrefix], index%workCodesPerPrefix)
}

func validWorkCodePrefix(prefix WorkCodePrefix) bool {
	for _, candidate := range workCodePrefixes {
		if prefix == candidate {
			return true
		}
	}
	return false
}
